src/DefaultBuilder/src/BootstrapHostBuilder.cs

```diff
@@ -98,36 +93,51 @@ public IHostBuilder ConfigureServices(Action<HostBuilderContext, IServiceCollect
             return this;
         }
 
-        public void RunDefaultCallbacks(HostBuilder innerBuilder)
+        public (HostBuilderContext, ConfigurationManager) RunDefaultCallbacks(HostBuilder innerBuilder)
         {
+            var hostConfiguration = new ConfigurationManager();
+
             foreach (var configureHostAction in _configureHostActions)
             {
-                configureHostAction(_configuration);
+                configureHostAction(hostConfiguration);
             }
 
             // This is the hosting environment based on configuration we've seen so far.
             var hostingEnvironment = new HostingEnvironment()
             {
-                ApplicationName = _configuration[HostDefaults.ApplicationKey],
-                EnvironmentName = _configuration[HostDefaults.EnvironmentKey] ?? Environments.Production,
-                ContentRootPath = HostingEnvironment.ResolveContentRootPath(_configuration[HostDefaults.ContentRootKey], AppContext.BaseDirectory),
+                ApplicationName = hostConfiguration[HostDefaults.ApplicationKey],
+                EnvironmentName = hostConfiguration[HostDefaults.EnvironmentKey] ?? Environments.Production,
+                ContentRootPath = HostingEnvironment.ResolveContentRootPath(hostConfiguration[HostDefaults.ContentRootKey], AppContext.BaseDirectory),
             };
 
             hostingEnvironment.ContentRootFileProvider = new PhysicalFileProvider(hostingEnvironment.ContentRootPath);
 
             var hostContext = new HostBuilderContext(Properties)
             {
-                Configuration = _configuration,
+                Configuration = hostConfiguration,
                 HostingEnvironment = hostingEnvironment,
             };
 
-            _configuration.SetBasePath(hostingEnvironment.ContentRootPath);
+            // Split the host configuration and app configuration so that the
+            // subsequent callback don't get a chance to modify the host configuration.
+            var configuration = new ConfigurationManager();
+            configuration.SetBasePath(hostingEnvironment.ContentRootPath);
+
+            // Chain the host configuration and app configuration together.
+            configuration.AddConfiguration(hostConfiguration, shouldDisposeConfiguration: true);
 
+            // ConfigureAppConfiguration sees the hosting configuration only, this is important
+            // so that code running during ConfigureAppConfiguration cannot see mutations to
+            // host configuration while running (e.g. the environment, content root, nor application name won't change).
```

Comment:
Maybe go with the following:

```suggestion
            // ConfigureAppConfiguration cannot modify the host configuration because doing so could
            // change the environment, content root and application name which is not allowed at this stage.
```